File: src/main/java/iteration2/RegistrationSystem.java
package iteration2;



import org.apache.log4j.LogManager;
import org.apache.log4j.Logger;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Random;


public class RegistrationSystem {
    private Department department = new Department();

    private JsonParser parser = new JsonParser();

    private Curriculum curriculum=new Curriculum();

    private Instructor instructor = new Instructor("dummy","dummy");

    private SystemParameter systemParameter;

    private static final Logger logger = LogManager.getLogger(RegistrationSystem.class);

    public RegistrationSystem() {

    }

    public Department getStudentManager() {
        return department;
    }

    public void setStudentManager(Department department) {
        this.department = department;
    }

    public JsonParser getParser() {
        return parser;
    }

    public void setParser(JsonParser parser) {
        this.parser = parser;
    }

    public Curriculum getCurriculum() {
        return curriculum;
    }

    public Instructor getInstructor() {
        return instructor;
    }

    public void setInstructor(Instructor instructor) {
        this.instructor = instructor;
    }

    public SystemParameter getSystemParameter() {
        return systemParameter;
    }

    public void setSystemParameter(SystemParameter systemParameter) {
        this.systemParameter = systemParameter;
    }

    public void readJsonFiles() throws IOException {
        this.systemParameter=this.parser.parseParameters();
        this.parser.parseAdvisors(this.department);
        Instructor [] instructors = this.department.getAdvisorList().toArray(new Instructor[0]);
        this.parser.parseCourseObjects(this.curriculum, instructors );
    }

    public void beginSimulation() throws IOException {
        this.initializeStudents();
        this.prepareInitializedStudents();
        logger.info("Starting simulation");
        this.parser.outputStudentObjects(
                this.department.getStudentList(),
                "./src/main/java/students/inputStudents/");
        enrollStudents();
        gradeStudents();
        calculateTranscript();
        this.parser.outputStudentObjects(
                this.department.getStudentList(),
                "./src/main/java/students/outputStudents/");
        logger.info("Ending simulation");
    }


    public void enrollStudents(){

        for (Student student : department.getStudentList()){
            ArrayList<Course> availableCourses = getAvailableCourses(student);
            student.enroll(availableCourses,curriculum,systemParameter);
        }
    }

    public void gradeStudents(){
        for (Instructor advisor : this.department.getAdvisorList().toArray(new Instructor[0])){
            for (Course course : advisor.getCoursesOfferedList()){
                for (String studentId : course.getEnrolledStudents()){
                    advisor.gradeStudents(department.getStudent(studentId),course);

                }
            }
        }
    }

    public void calculateTranscript(){
        for (Student student : department.getStudentList()){
            student.getTranscript().getSemesters().add(student.getStudentSemester());
            for (StudentSemester semester : student.getTranscript().getSemesters()){
                semester.calculateYano();
                semester.setLetterGrades(new ArrayList<>());
                semester.calculateLetterGrade();
            }
            student.getTranscript().calculateGpa();
        }
    }

    public ArrayList<Course> getAvailableCourses(Student student) {
        ArrayList<Course>[] courses = getCurriculum().getCOURSES();
        ArrayList<Course> availableCourses = new ArrayList<>();
        for (ArrayList<Course> courseList : courses){
            for (Course course : courseList){
                if (student.getTranscript().containsCourse(course.getCode())){
                    if(student.getTranscript().getMaxGrade(course.getCode()) < 2){
                        availableCourses.add(course);
                    }
                }
                else{
                    availableCourses.add(course);
                }
            }
        }
        return availableCourses;
    }

    public void initializeStudents(){
        for (int i=1;i<=4;i++){
            int term = 2 * (i-1) + this.systemParameter.getSemester();
            String departmentCode= "1501";
            String entryYear = Integer.toString(22 - i);
            for (int j=1;j<=this.systemParameter.getStudentPerSemester();j++){
                String entryPlace = j < 10 ? "00"+j : "0"+j ;
                String studentNumber = departmentCode + entryYear + entryPlace;
                Student student = new Student(studentNumber,studentNumber, term);
                Advisor advisor = department.getAdvisorList().get(new Random().nextInt(department.getAdvisorList().size()));
                student.setAdvisor(advisor);
                logger.info("Added advisor" + advisor.getName() + " to student " + student.getName() );
                department.getStudentList().add(student);
            }
        }
    }

    public void prepareInitializedStudents() throws IOException {
        for (int i=1;i<=4;i++){
            int currentStudentTerm = 2 * (i -1) + this.systemParameter.getSemester();
            for (int k=1;k<currentStudentTerm;k++){
                for (int j=0;j<this.systemParameter.getStudentPerSemester();j++){
                    Student student =this.department.getStudentList()
                            .get( (i-1) * this.systemParameter.getStudentPerSemester() + j );
                    student.setTerm(k);
                    student.setStudentSemester(new StudentSemester(k));
                    student.setEnrolledCourses(new ArrayList<String>());
                    ArrayList<Course> availableCourses = getAvailableCourses(student);
                    student.enroll(availableCourses,curriculum,systemParameter);
                    //grade student
                    this.gradeStudents();

                    student.getStudentSemester().calculateLetterGrade();
                    student.getStudentSemester().calculateYano();
                    student.getTranscript().getSemesters().add(student.getStudentSemester());
                    student.getTranscript().calculateGpa();
                    student.setTerm(currentStudentTerm);

                    // resetting curriculum
                    this.curriculum = new Curriculum();
                    Instructor[] instructors = this.parser.parseAdvisors(this.department).toArray(new Instructor[0]);
                    this.parser.parseCourseObjects(this.curriculum,instructors);
                }
                this.curriculum = new Curriculum();
                this.parser.parseCourseObjects(this.curriculum,this.department.getAdvisorList().toArray(new Advisor[0]));
            }
        }
        for (Student student : this.department.getStudentList()){
            student.setStudentSemester(new StudentSemester(student.getTerm()));
            student.setEnrolledCourses(new ArrayList<>());
            student.setLogs(new ArrayList<>());
            System.out.println("Student is : " + student.toString());
        }
        this.curriculum = new Curriculum();
        // TODO
        // CREATE NEW INSTRUCTORS
        Instructor[] instructors = this.parser.parseAdvisors(this.department).toArray(new Instructor[0]);
        this.parser.parseCourseObjects(this.curriculum,instructors);
    }

    public void setCurriculum(Curriculum curriculum) {
        this.curriculum = curriculum;
    }
}

File: RegistrationSystem.py
import random

from Student import Student
from StudentSemester import StudentSemester
from StudentSemester import StudentSemester
from Curriculum import Curriculum
from JsonParser import JsonParser
from Department import Department
from SystemParameter import SystemParameter


class RegistrationSystem:

    def __init__(self):
        self.jsonParser = JsonParser()
        self.curriculum = Curriculum()
        self.department = Department()
        self.systemParameter = self.jsonParser.parseParameters()

    def readJsonFiles(self):
        self.jsonParser.parseAdvisors(self.department)
        self.jsonParser.parseCourses(self.curriculum, self.department.get_advisor_list())

    def beginSimulation(self):
        self.initializeStudents()
        self.prepareInitializedStudents()
        print("Simulation starting")
        self.jsonParser.outputStudentObjects(self.department.get_student_list(), "./students/inputStudents")
        self.enrollStudents()
        self.gradeStudents()
        self.calculateTranscript()
        self.jsonParser.outputStudentObjects(self.department.get_student_list(), "C:/Users/Mustafa/Desktop/iteson/CSE3063F22P1_GRP4/students/outputStudents")
        print("Simulation ending")

    def enrollStudents(self):
        for student in self.department.student_list:
            availableCourses = self.getAvailableCourses(student)
            student.enroll(availableCourses, self.curriculum, self.systemParameter)

    def calculateTranscript(self):
        for student in self.department.get_student_list():
            student.transcript.get_semesters().append(student.student_semester)
            for std_semester in student.transcript.get_semesters():
                std_semester.calculate_yano()
                std_semester.letter_grades = []
                std_semester.calculate_letter_grade()
            student.transcript.calculate_gpa()

    def gradeStudents(self):
        for advisor in self.department.get_advisor_list():
            for course in advisor.courses_offered_list:
                for studentId in course.enrolledStudents:
                    advisor.grade_students(self.department.get_student(studentId), course)

    def getAvailableCourses(self, student: Student):
        courses = self.curriculum.courses
        availableCourses = []
        for courseList in courses:
            for course in courseList:
                if student.transcript.contains_course(course.code):
                    if student.transcript.get_max_grade(course.code) < 2:
                        availableCourses.append(course)
                else:
                    availableCourses.append(course)

        return availableCourses

    def initializeStudents(self):
        for i in range(1, 5):
            term = 2 * (i - 1) + self.systemParameter.semester
            departmentCode = "1501"
            entryYear = str(22 - i)
            for j in range(1, self.systemParameter.student_per_semester + 1):
                entryPlace = ""
                if j < 10:
                    entryPlace = "00" + str(j)
                else:
                    entryPlace = "0" + str(j)
                studentNumber = departmentCode + entryYear + entryPlace
                student = Student(studentNumber, studentNumber, term)
                advisor = self.department.get_advisor_list()[
                    random.randint(0, len(self.department.get_advisor_list()) - 1)]
                student.advisor = advisor
                self.department.student_list.append(student)

    def prepareInitializedStudents(self):
        for i in range(1, 5):
            currentStudentTerm = 2 * (i - 1) + self.systemParameter.semester
            for k in range(1, currentStudentTerm):
                for j in range(0, self.systemParameter.student_per_semester):
                    student = self.department.get_student_list()[
                        (i - 1) * self.systemParameter.student_per_semester + j]
                    student.term = k
                    student.student_semester = StudentSemester(k)
                    student.enrolled_courses = []
                    availableCourses = self.getAvailableCourses(student)
                    student.enroll(availableCourses, self.curriculum, self.systemParameter)

                    self.gradeStudents()
                    student.student_semester.calculate_letter_grade()
                    student.student_semester.calculate_yano()
                    student.transcript.get_semesters().append(student.student_semester)
                    student.transcript.calculate_gpa()
                    student.term = currentStudentTerm

                    # resetting curriculum
                    self.curriculum = Curriculum()
                    self.department.advisor_list = []
                    self.jsonParser.parseAdvisors(self.department)
                    self.jsonParser.parseCourses(self.curriculum, self.department.get_advisor_list())

            self.curriculum = Curriculum()
            self.jsonParser.parseCourses(self.curriculum, self.department.get_advisor_list())

        for student in self.department.get_student_list():
            student.student_semester = StudentSemester(student.term)
            student.enrolled_courses = []
            student.logs = []

        self.curriculum = Curriculum()
        self.department.advisor_list = []
        self.jsonParser.parseAdvisors(self.department)
